metadata-ingestion/src/datahub/ingestion/source/tableau.py

```diff
@@ -973,11 +1178,44 @@ def get_schema_metadata_for_custom_sql(
         )
         return schema_metadata
 
-    def _get_project(self, node):
-        if node.get("__typename") == "EmbeddedDatasource" and node.get("workbook"):
-            return node["workbook"].get("projectName")
-        elif node.get("__typename") == "PublishedDatasource":
-            return node.get("projectName")
+    def _get_project(self, ds):
+        # For PublishedDatasource the GraphQL API doesn't return the projectLuid and hence we need to fetch
+        # the projectLuid from datasource_project_map
+        # For EmbeddedDatasource the GraphQL API returns the projectLuid in resposne and
+        # hence it is available at ds["workbook"]["projectLuid"]
+        if self.config.extract_project_hierarchy:
+            if (
+                ds.get("__typename") == "PublishedDatasource"
+                and ds.get("luid")
+                and ds["luid"] in self.datasource_project_map.keys()
+                and self.datasource_project_map[ds["luid"]]
+                in self.tableau_project_registry.keys()
+            ):
+                return self._get_project_browse_path(
+                    self.datasource_project_map[ds["luid"]]
+                )
+            elif (
+                ds.get("__typename") == "EmbeddedDatasource"
+                and ds.get("workbook")
+                and ds.get("workbook").get("projectLuid")
+                and ds["workbook"]["projectLuid"]
+                in self.tableau_project_registry.keys()
+            ):
+                return self._get_project_browse_path(ds["workbook"]["projectLuid"])
+
+            datasource_name: str = ds.get("name")
+            logger.warning(
+                f"Could not load project hierarchy for datasource {datasource_name}. Please check permissions."
+            )
+            logger.debug(f"datasource = {ds}")
+
+            return None
+
+        # default behavior in absence of extract_project_hierarchy
+        elif ds.get("__typename") == "EmbeddedDatasource" and ds.get("workbook"):
+            return ds["workbook"].get("projectName")
```

Comment:
why do we do this part? 

why don't we just always provide the correct project name back? seems tome like over-engineering

Reply:
like why do we need to gate behind "extract_project_hierarchy" - this part doesn't make sense to me